hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion6Strategy.java

```diff
@@ -98,9 +98,9 @@ public UUID generateUuid(SharedSessionContractImplementor session) {
 	private long getSequence(final long currentTimestamp) {
 		lock.lock();
```

Comment:
I'm not an expert on UUID, but it seems to me that this could be implemented in a lock-free manner by making `currentTimestamp` an `AtomicLong` like this:

```java
		final long previousTimestamp = this.currentTimestamp.get();
		if ( previousTimestamp <= currentTimestamp ) {
			if ( this.currentTimestamp.compareAndSet( previousTimestamp, currentTimestamp ) ) {
				final byte[] bytes = new byte[2];
				Holder.numberGenerator.nextBytes( bytes );
				final int randomShort = ( ( bytes[0] & 0xFF ) << 8 ) | ( bytes[1] & 0xFF );
				// Use the lower 14 bits
				final int clockSequence = randomShort & 0x2FFF;
				this.clockSequence.set( clockSequence );
				return clockSequence;
			}
		}
		return this.clockSequence.updateAndGet( i -> ( i + 1 ) & 0x2FFF );
```

Reply:
This has not much with UUID itself, only I've tried to be sure that those sequence numbers will be created in equested order. Maybe this is too conservative approach, I'm not 100% sure.

Reply:
I've changed code so that now both timestamp and sequence are kept as immutable object. This will allow both to be atomically calculated and changed at the same time. One of "side effects" is that now all locks are redundant. At the same time, this will simplify any change in logic.